### PR TITLE
Hot events for Informer API 

### PR DESCRIPTION
**Pull request by MSNTCS**
Informer API is supposed to send events as soon as they are triggered to all subscribed clients. In this PR, Hot events, those are not retrieved from the history of the event, are covered.

### CODE REVIEW

File: informer/src/handler/connection.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+impl Connection{
+    pub fn new(status: bool, events: String , sink: Sink) -> Self {
+
+        Self{
+            status: status,
+            intrested_event: vec![events.clone()],
+            sink: sink.clone(),
+        }
+    }
+    pub fn event_subscription <T> (&mut self, mut io : PubSubHandler<T>) {
```
Comment (majecty) [inline informer/src/handler/connection.rs]: Will you add this function in the Connection struct? I feel awkward. The connection will be created whenever a user is connected. `io.add_subscription` will be called only once.

Comment (MSNTCS) [reply informer/src/handler/connection.rs]: I think `io.add_subscription` will be called whenever a user sends a subscription request. and will not be called once.

Comment (majecty) [reply informer/src/handler/connection.rs]: Since their documentation is not good, it is hard to know their intention.
With my common sense about webserver, what `io.add_subscription` is doing is registering a callback to the WebSocket server. In web server, we do not register callback multiple times.

In my assumption, the closure you passed to the `io.add_subscription`' will be called whenever user connects.

Could you check the `io.add_subscription`'s behavior?

File: informer/src/handler/connection.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                    .reject(Error{
+                        code: ErrorCode::ParseError,
+                        message: "Invalid parameters. Subscription rejected.".into(),
+                        data: None,
+                    })
+                    .unwrap();
+                return;
+            }
+
+            thread::spawn(move || {
```
Comment (majecty) [inline informer/src/handler/connection.rs]: Why do we need a new thread here?

Comment (MSNTCS) [reply informer/src/handler/connection.rs]: This is for my experience of how we can manage a thread. do not mind it for now.

File: informer/src/handler/connection.rs (hunk 1 above)
Comment (MSNTCS) [reply informer/src/handler/connection.rs]: `add_subscription` creates a new `subscribe` and `unsubscribe` RPC methods which their types are `SubscribeRpcMethod` and `UnsubscribeRpcMethod`. Then add these two so-called asynchronous methods to RPC request handler (`MetaIoHandler`). Therefore, I cannot conclude whether we need to create a subscription and unsubscription for each event by ourself (before anybody sends any subscription request) or we need to create a subscription/un per each request.

File: informer/src/handler/connection.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{Events, Params, Sink, Value};
+use futures::Future;
+use jsonrpc_core::futures;
+
+#[derive(Clone)]
```
Comment (majecty) [inline informer/src/handler/connection.rs]: When do you need to clone the Connection struct?

File: informer/src/informer_service/dummy_informer_service.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+#[derive(Clone, PartialEq)]
+pub enum Events {
+    None,
+    PeerAdded(String, String, usize),
+}
+#[derive(Clone)]
+pub struct InformerService {
+    connections: Vec<Connection>,
+    pub sender: NotifySender,
```
Comment (majecty) [inline informer/src/informer_service/dummy_informer_service.rs]: Why does the InformerService have a sender? I thought that the InformerService would have only receivers.

Comment (MSNTCS) [reply informer/src/informer_service/dummy_informer_service.rs]: Because I create a channel for sending connection in  `run node` and share with `handler` and `informer service`. There

Comment (majecty) [reply informer/src/informer_service/dummy_informer_service.rs]: IMO, since the sender is not used in the InformerService, it should not have it.
There are two options for how to remove the sender from the InfomreService struct.

One: making InformerService::new return the sender. For example
```rs
pub fn new(connection_receiver: Receiver<Connection>) -> (Self, NotifySender) {
```

Two: making the InformerService::new receive the sender. For example:
```rs
    pub fn new(connection_receiver: Receiver<Connection>, event_receiver: Receiver<Events>) -> Self {
```

Comment (MSNTCS) [reply informer/src/informer_service/dummy_informer_service.rs]: I got your point and am agree with you. I will use the first way.

File: informer/src/handler/connection.rs (hunk 3 above)
Comment (MSNTCS) [reply informer/src/handler/connection.rs]: I need to clone the connections of InfromerService when the code is notifying all clients.

File: informer/src/handler/rpc_ws_handler.rs [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    }
+    pub fn event_subscription(&mut self, sender: Sender<Connection>) {
+        self.handler.add_subscription(
+            "Your event subscription begins",
+            ("register", move |params: Params, _, subscriber: Subscriber| {
+                let paramsclone = params.clone();
+                let unwraped_params: Result<Vec<String>, Error> = params.parse();
+                let allparams = unwraped_params.unwrap();
+                println!("####################");
```
Comment (MSNTCS) [inline informer/src/handler/rpc_ws_handler.rs]: @majecty I forgot to change these, I will change it after your review is done.

File: informer/src/handler/connection.rs [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                let event_params: Vec<Value> = vec![message.into(), address.into(), token.into()];
+                match self.sink.notify(Params::Array(event_params)).wait() {
+                    Ok(_) => {}
+                    Err(_) => {
+                        cinfo!(INFORMER, "Subscription has ended, finishing.");
+                    }
+                }
+            }
+            _ => cinfo!(INFORMER, "invalid Event: the event is not supported"),
+        }
```
Comment (majecty) [inline informer/src/handler/connection.rs]: You can make the code simple using serde.

```
let json_object = serde_json::to_value(event).unwrap().as_object();
let params = Params::Map(json_object);
match self.sink.notify(Params::Array(event_params)).wait() {
    Ok(_) => {}
    Err(_) => {
        cinfo!(INFORMER, "Subscription has ended, finishing.");
    }   
}
```

File: informer/src/handler/rpc_ws_handler.rs [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+use std::io;
+use std::sync::Arc;
+
+pub struct InformerConfig {
+    pub interface: String,
+    pub port: u16,
+    pub max_connections: usize,
+}
+
+#[derive(Default)]
```
Comment (majecty) [inline informer/src/handler/rpc_ws_handler.rs]: Why do you need `Default`?

File: informer/src/handler/rpc_ws_handler.rs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    }
+    pub fn start_ws(self, config: InformerConfig) -> Result<WsServer, String> {
+        let url = format!("{}:{}", config.interface, config.port);
+        let addr = url.parse().map_err(|_| format!("Invalid WebSockets listen host/port given: {}", url))?;
+        let server = self.handler;
+        let start_result = start_ws(&addr, server, config.max_connections);
+        match start_result {
+            Err(WsError::Io(ref err)) if err.kind() == io::ErrorKind::AddrInUse => {
+                Err(format!("WebSockets address {} is already in use, make sure that another instance of a Codechain node is not running or change the address using the --ws-port options.", addr))
```
Comment (majecty) [inline informer/src/handler/rpc_ws_handler.rs]: Is informer using `--ws-port` option?

File: informer/src/handler/rpc_ws_handler.rs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            },
+            Ok(server) => {
+                cinfo!(INFORMER, "WebSockets Listening on {}", addr);
+                Ok(server)
+            },
+        }
+    }
+    pub fn event_subscription(&mut self, sender: Sender<Connection>) {
+        self.handler.add_subscription(
+            "Your event subscription begins",
```
Comment (majecty) [inline informer/src/handler/rpc_ws_handler.rs]: What message should we use here?

File: informer/src/handler/rpc_ws_handler.rs [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+            ("register", move |params: Params, _, subscriber: Subscriber| {
+                let paramsclone = params.clone();
+                let unwraped_params: Result<Vec<String>, Error> = params.parse();
+                let allparams = unwraped_params.unwrap();
```
Comment (majecty) [inline informer/src/handler/rpc_ws_handler.rs]: You should not `unwrap` here. When a user gives wrong parameter, Foundry should not panic.

File: informer/src/handler/rpc_ws_handler.rs [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+                cinfo!(INFORMER, "WebSockets Listening on {}", addr);
+                Ok(server)
+            },
+        }
+    }
+    pub fn event_subscription(&mut self, sender: Sender<Connection>) {
+        self.handler.add_subscription(
+            "Your event subscription begins",
+            ("register", move |params: Params, _, subscriber: Subscriber| {
+                let paramsclone = params.clone();
```
Comment (majecty) [inline informer/src/handler/rpc_ws_handler.rs]: Why did you clone params here?

File: informer/src/handler/rpc_ws_handler.rs [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+                println!("####################");
+                println!("{}", allparams[0]);
+                if paramsclone == Params::None {
+                    subscriber
+                        .reject(Error {
+                            code: ErrorCode::ParseError,
+                            message: "Invalid parameters. Subscription rejected.".into(),
+                            data: None,
+                        })
+                        .unwrap();
```
Comment (majecty) [inline informer/src/handler/rpc_ws_handler.rs]: Do you know when the `unwrap` fails? Please use `expect` rather than `unwrap` and use a message that contains why you convinced that this will not happen.

File: informer/src/handler/rpc_ws_handler.rs [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                let mut rng = rand::thread_rng();
+                let sub_id = rng.gen();
+                println!("{}", sub_id);
+                let sink = subscriber.assign_id(SubscriptionId::Number(sub_id)).unwrap();
+                match sink.notify(Params::Array(vec![Value::String("You are now a new subscriber".into())])).wait() {
+                    Ok(_) => {}
+                    Err(_) => {
+                        cinfo!(INFORMER, "Subscription has ended, finishing.");
+                    }
+                }
```
Comment (majecty) [inline informer/src/handler/rpc_ws_handler.rs]: Is this needed?

File: foundry/config/mod.rs [hunk 14]
```diff
@@ -125,6 +128,16 @@ impl Config {
         }
     }
 
+    pub fn informer_config(&self) -> InformerConfig {
+        debug_assert!(!self.informer.disable.unwrap());
```
[the remark this answers is not in the document]
Comment (MSNTCS) [reply foundry/config/mod.rs]: I followed the same pattern as other functions.

```
pub fn rpc_ipc_config(&self) -> RpcIpcConfig {
        debug_assert!(!self.ipc.disable.unwrap());

        RpcIpcConfig {
            socket_addr: self.ipc.path.clone().unwrap(),
        }
    }
```

File: informer/src/handler/rpc_ws_handler.rs (hunk 9 above)
Comment (MSNTCS) [reply informer/src/handler/rpc_ws_handler.rs]: here is what a user received.
```
Received:  {"jsonrpc":"2.0","method":"Your event subscription begins","params":["You are now a new subscriber"]}
```

File: informer/src/handler/rpc_ws_handler.rs (hunk 13 above)
Comment (MSNTCS) [reply informer/src/handler/rpc_ws_handler.rs]: The purpose of this message is to let the client know that the subscription successfully occurs.
here is what the client receives after a successful subscription.
```
Received:  {"jsonrpc":"2.0","result":14308347713743523176,"id":1}
Received:  {"jsonrpc":"2.0","method":"Your event subscription begins","params":["You are now a new subscriber"]}
```

File: foundry/run_node.rs [hunk 15]
```diff
@@ -317,6 +348,18 @@ pub fn run_node(matches: &ArgMatches<'_>) -> Result<(), String> {
         self_nominate_start(c, matches, accountp, address);
     }
 
+    let mut _informer_handler = Handler::default();
```
Comment (majecty) [inline foundry/run_node.rs]: Why did you add prefix "_" in the variable name?

File: foundry/run_node.rs [hunk 16]
```diff
@@ -317,6 +348,18 @@ pub fn run_node(matches: &ArgMatches<'_>) -> Result<(), String> {
...
+    let _informer_server = {
```
Comment (majecty) [inline foundry/run_node.rs]: ditto.

**Comment by majecty** [comment]
Please use `expect` instead of `unwrap`.
We can use `unwrap` only when the situation is obvious.

File: informer/src/handler/rpc_ws_handler.rs (hunk 9 above)
Comment (majecty) [reply informer/src/handler/rpc_ws_handler.rs]: We don't need it. Do you know how to prohibit the library from sending the message?

File: informer/src/informer_service/informer_notify.rs [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+use crate::Events;
+use crossbeam::{bounded, unbounded, Receiver, Sender};
+use crossbeam_channel as crossbeam;
+
+pub fn create() -> (NotifySender, Receiver<Events>) {
+    let (tx, rx) = unbounded();
```
[the remark this answers is not in the document]
Comment (MSNTCS) [reply informer/src/informer_service/informer_notify.rs]: I add comments

File: informer/src/handler/rpc_ws_handler.rs [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+                let unwrapped_params: Result<Vec<String>, Error> = params.parse();
+                let all_params =
+                    unwrapped_params.expect("The parameter that the client sent for subscription is not valid");
+                let mut rng = rand::thread_rng();
+                let sub_id = rng.gen();
+                let sink = subscriber
+                    .assign_id(SubscriptionId::Number(sub_id))
+                    .expect("Subscription request is already terminated");
+                let mut connection = Connection::new(true, sink, sub_id);
```
Comment (majecty) [inline informer/src/handler/rpc_ws_handler.rs]: It's hard to know what it is when you just pass the `true` value.
I have two suggestions.
* Use an enum for it like "ConnectionState::Connected" and "ConnectionState::Disconnected".
* Do not receive the state from the `new` function. You will do not want to create a connection with the state value `false`.

File: informer/src/informer_service/informer_service_handler.rs [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+            cinfo!(INFORMER, "Informer Service has started");
+            let mut select = crossbeam::Select::new();
+            let event_index = select.recv(&event_rcv);
+            let connection_index = select.recv(&connection_rcv);
+            loop {
+                let mut s = select.clone();
+                match s.ready() {
+                    index if index == event_index => {
+                        if let Ok(event) = event_rcv.try_recv() {
+                            cinfo!(INFORMER, "Event is send to all clients");
```
Comment (majecty) [inline informer/src/informer_service/informer_service_handler.rs]:
```suggestion
                            cinfo!(INFORMER, "Event is sent to all clients");
```

File: informer/src/informer_service/informer_service_handler.rs [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+                            let popup_event = event;
```
Comment (majecty) [inline informer/src/informer_service/informer_service_handler.rs]: Why did you create a new variable?

File: informer/src/handler/connection.rs [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+                cinfo!(INFORMER, "The event is successfully added to user's interested events");
+                self.intrested_events.push(event);
+            }
+            _ => {
+                cinfo!(INFORMER, "invalid Event: the event is not supported");
+            }
+        }
+    }
+
+    pub fn notify_client(&mut self, event: &Events) {
```
Comment (majecty) [inline informer/src/handler/connection.rs]: You can remove `mut` from the `notify_client` function like this:
```
diff --git a/informer/src/handler/connection.rs b/informer/src/handler/connection.rs
index d2b678e4..8d6c085e 100644
--- a/informer/src/handler/connection.rs
+++ b/informer/src/handler/connection.rs
@@ -47,7 +47,7 @@ impl Connection {
         }
     }
 
-    pub fn notify_client(&mut self, event: &Events) {
+    pub fn notify_client(&self, event: &Events) {
         let json_object = serde_json::to_value(event).unwrap().as_object_mut().cloned();
         let params = Params::Map(json_object.expect("json format is not valid"));
         match self.sink.notify(params).wait() {
diff --git a/informer/src/informer_service/informer_service_handler.rs b/informer/src/informer_service/informer_service_handler.rs
index e40e3aa2..32f8a3c2 100644
--- a/informer/src/informer_service/informer_service_handler.rs
+++ b/informer/src/informer_service/informer_service_handler.rs
@@ -90,8 +90,8 @@ impl InformerService {
         }
     }
 
-    pub fn notify_client(&mut self, popup_event: Events) {
-        for connection in &mut self.connections {
+    pub fn notify_client(&self, popup_event: Events) {
+        for connection in &self.connections {
             for interested_event in connection.intrested_events.clone() {
                 if InformerService::compare_event_types(&interested_event, &popup_event) {
                     connection.notify_client(&popup_event);

```

File: informer/src/informer_service/informer_service_handler.rs [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+        let event_rcv = self.event_receiver.clone();
+        let connection_rcv = self.connection_receiver.clone();
+        spawn(move || {
+            cinfo!(INFORMER, "Informer Service has started");
+            let mut select = crossbeam::Select::new();
+            let event_index = select.recv(&event_rcv);
+            let connection_index = select.recv(&connection_rcv);
+            loop {
+                let mut s = select.clone();
+                match s.ready() {
```
Comment (majecty) [inline informer/src/informer_service/informer_service_handler.rs]:
```suggestion
                match select.ready() {
```

File: network/src/p2p/handler.rs [hunk 23]
```diff
@@ -120,8 +121,11 @@ pub struct Handler {
     max_peers: usize,
     peer_db: Box<dyn (ManagingPeerdb)>,
     rng: Mutex<OsRng>,
+
+    event_sender: NotifySender,
```
Comment (majecty) [inline network/src/p2p/handler.rs]: The names`NotifySender` and `event_sender` are too general.
It is hrad to know what they are using the name.
How about `informer_sender: InformerSender` or `informer_event_sender: InformerEventSender`?

File: informer/src/handler/rpc_ws_handler.rs [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+                            code: ErrorCode::ParseError,
+                            message: "Invalid parameters. Subscription rejected.".into(),
+                            data: None,
+                        })
+                        .expect("Invalid parameters are used. Subscription rejected");
+                    return
+                }
+                let unwrapped_params: Result<Vec<String>, Error> = params.parse();
+                let all_params =
+                    unwrapped_params.expect("The parameter that the client sent for subscription is not valid");
```
[the remark this answers is not in the document]
Comment (MSNTCS) [reply informer/src/handler/rpc_ws_handler.rs]: This error would not be for the client. this error is for foundry.
